### ISSUE TITLE
Problem rendering data set in will-mount

### ISSUE THREAD

**Issue by bilus** [omcljs/om#336]
Hi,

I've upgraded to the latest version of om (0.8.8) and the following snippet no longer works:

```clojure
(def app-state (atom {}))

(defn sample
  [app-state owner]
  (reify
    om/IWillMount
    (will-mount [_]
                (om/transact! app-state (fn [app-state]
                                          (assoc-in app-state [:whatever] "something"))))
    om/IRender
    (render [_]
            (dom/div {} (:whatever app-state))))

(defn main []
  (om/root
    sample
    app-state
    {:target (. js/document (getElementById "app"))}))
```

(it renders nothing after a browser refresh)

when you change the `render` function to what you see below, it works correctly:

```clojure
(render [_]
            (dom/div {} (:whatever @app-state)))
```

If you have trouble reproducing it, please let me know; the project has additional dependencies and while I've tried to distill it as much as possible, things happen.

**Comment by mtbrooks33** [comment]
Try changing the `app-state` symbol passed into the component to `app`. For example, `(defn sample [app owner] ...)`. Clojurescript might be getting confused with the `(def app-state ...)` and the `app-state` passed into the function. The fact that it works then derefrencing `app-state` with `@` makes this even more apparent.

**Comment by bilus** [comment]
O, wow, will give it a try, thank you.

**Comment by JulianLeviston** [comment]
Perhaps this is spreading fear, uncertainty and doubt.

Clojurescript won't "get confused" by naming an argument to a function the same as something else within a namespace. The only way this would be a problem is if you need to access the symbol your argument is shadowing.

Having said that, the code in question can easily be reduced to (om/transact! app-state assoc-in [:whatever] "something")

Cursors have changed a bit since 0.8, though, so it might be quite useful to:
1. Close this
2. Re-read the intro to Om, and do the tutorial again, or at least skim through it.

**Comment by bilus** [comment]
I'm glad to hear this isn't a cljs problem. Could you tell me what is wrong with the code below? I'd really appreciate that.

Based on the basic tutorial:

```clojure
(defonce app-state (atom {:text "original value"}))

(defn main []
  (om/root
    (fn [data owner]
      (reify
        om/IWillMount
        (will-mount [_]
          (om/transact! data (fn [data]
                               (assoc-in data [:text] "something"))))
        om/IRender
        (render [_]
          (dom/p nil (:text data)))))
    app-state
    {:target (. js/document (getElementById "app"))}))
```

It renders "original value" instead of "something" (using om 0.8.8 which I think is the latest one).

**Comment by bilus** [comment]
Also, this may help in recreating this issue `[org.clojure/clojurescript "0.0-2913"]`(unless it's a case of my brain's malfunction).

**Comment by bensu** [comment]
I took this as an excuse to dig into `om.core/core.cljs`. This is what is going on:

1. Om calls `React.render` on your component with `"original value"`.
2. While mounting, you `transact` to `"something"`.
3. That triggers a watch that adds it to the new state and then decides if it should trigger a new render.
4. The watch decides that it [should render](https://github.com/omcljs/om/blob/master/src/om/core.cljs#L1165) because the state changed.
5. The watch gives back control to the component which now calls `render` with `"original value"`.
6. When the rendering is finished, since it just rendered it decides it [shouldn't render again](https://github.com/omcljs/om/blob/master/src/om/core.cljs#L1159), undoing step 4.
6. You don't see what you wanted and no future renders are called.

With the following example we can see two things: that the transacted state is preserved and that `transact` doesn't trigger new renders during the Lifecycle:

```clj
(defonce app-state (atom {:text "original value"
                          :other true}))

(om/root
 (fn [data owner]
   (reify
     om/IWillMount
     (will-mount [_]
       (om/transact! data (fn [data]
                            (assoc-in data [:text] "something"))))
     om/IDidMount
     (did-mount [_]
       (om/transact! data (fn [data]
                            (assoc-in data [:text] "something else"))))
     om/IRender
     (render [_]
       (dom/div nil
                (dom/p nil (:text data))
                (dom/button #js {:onClick (fn [_]
                                            (om/transact! data :other not))}
                            "Toggle")))))
 app-state
 {:target (. js/document (getElementById "app"))})
```

If you click the button, it will render the last value: `"something else"`. 

The question is then, should `transact` schedule a new render inside the Lifecycle methods? Since it worked before, it might have been a bug then and a feature now. Or backwards. That is for the author to determine :)

In case it is a bug, a solution would be to decide if the [next render should be skipped](https://github.com/omcljs/om/blob/master/src/om/core.cljs#L1159) right [before](https://github.com/omcljs/om/blob/master/src/om/core.cljs#L1130) the current render. I can test it with the examples and submit a PR if needed.

In case you really want to do this, delay your transaction with an ugly timeout:

```clj
     (will-mount [_]
       (go (let [_ (<! (timeout 1))]
             (om/transact! data (fn [data]
                                  (assoc-in data [:text] "something"))))))
```

**Comment by cedrictools** [comment]
The :fn option of om/build achieve somehow similar behavior to transact in will-mount isn't it ?
https://github.com/omcljs/om/wiki/Documentation#build
:fn - a function to apply to x before invoking f.

**Comment by bensu** [comment]
@cedrictools no, `:fn` will be applied on every render while things you during `will-mount` will only happen once. If you chose `fn` so that it transforms `(:text data)` into `"something"` you'll always see `"something"` on the screen.

The objective of `:fn` is to adapt the data in some way so that the component understands it. For example, wrap it in a vector. It should have no side-effects.

For example, my cursor has `["one","two","three"]`, but my component expects `[{:text "one"},{:text "two"},{:text "three}]`. Then I would use `:fn #(mapv (fn [x] {:text x}) %)`.

**Comment by cedrictools** [comment]
You're right, on every render, missed that and it's really different. Thanks for clarifying that point and sorry for the noise.

**Comment by bilus** [comment]
@bensu Great work! I do have a workaround, thanks, but I didn't think that was an intuitive behaviour. 

The way it works has certainly changed a lot since 0.7.x which was when I last read the sources that's why it confused me but I still don't think you should have to use timeouts to do `transact!` in `will-mount`. The case is a wizard control that initializes its state in `will-mount`:

```clojure
(defn wizard
  [steps data-korks wizard-korks & {:keys [before success cancel]}]
  (fn [app-state owner]
    (let [initial-step (first steps)
          wizard-cursor (get-in app-state wizard-korks)
          data-cursor (get-in app-state data-korks)]
      (reify
        om/IWillMount
        (will-mount [_]
          (event-bus/bubble owner {:event ::mount})
          (when before
            (before owner [data-cursor wizard-cursor]))
          (om/transact! wizard-cursor (fn [wizard]
                                 (-> wizard
                                     (assoc-in [:step-history] (list initial-step))
                                     (assoc-in [:all-steps] steps)
                                     (assoc-in [:on-cancel] #(do (when cancel (cancel owner [data-cursor wizard-cursor]))
                                                                 (event-bus/bubble owner {:event ::cancel})))
                                     (assoc-in [:on-success] #(do (when success (success owner [data-cursor wizard-cursor]))
                                                                  (event-bus/bubble owner {:event ::success})))))))
        om/IRender
        (render [_]
          (om/build (or (current-step @wizard-cursor)       ;; https://github.com/omcljs/om/issues/336
                        wait-view)
```

Maybe there's a better way to do it (there are *always* better ways to do things:) but in any case I think that having to understand the, quite complicated, innards of om to do a pretty straightforward thing is a bit much.

TL;DR +1 PR ;)

**Comment by bensu** [comment]
Hi @bilus I think this problem was solved by #342

**Comment by bilus** [comment]
Thanks a lot, will have a look.

**Comment by swannodette** [comment]
Closing this one unless for spring cleaning. If it's still a problem reopen.
